jupyterhub_moss/spawner.py

```diff
@@ -79,11 +82,80 @@ def _validate_partitions(self, proposal):
         return partitions
 
     slurm_info_cmd = traitlets.Unicode(
-        # Get number of nodes and cores for all partitions
-        r"sinfo -a -N --noheader -o '%R %t %m'",
-        help="Command to query cluster information from Slurm. Formatted using req_xyz traits as {xyz}.",
+        # Get number of nodes, cores, gpus, total memory for all partitions
+        r"sinfo -a --noheader -o '%R %D %C %G %m'",
+        help="Command to query cluster information from Slurm. Formatted using req_xyz traits as {xyz}."
+        "Output will be parsed by ``slurm_info_resources``.",
     ).tag(config=True)
 
+    slurm_info_resources = traitlets.Callable(
+        help="""Provides information about resources in Slurm cluster.
+
+        It will be called with the output of ``slurm_info_cmd`` as argument and should return a tuple:
+        - list of resource labels to be displayed in table of available resources
+        - dictionary mapping each partition name to resources defined in ``RESOURCES_COUNTS``""",
+    ).tag(config=True)
+
+    @traitlets.default("slurm_info_resources")
+    def _slurm_info_resources(self, slurm_info_out):
+        """
+        Parses output from Slurm command: sinfo -a --noheader -o '%R %D %C %G %m'
+        Returns information about partition resources listed in RESOURCES_COUNTS: number of cores,
+        max memory, gpus and resource counts to be shown in table of available resources
+        :param slurm_info_out: string with output of slurm_info_cmd
+        :rtype: tuple with:
+            - list of resource labels to be displayed in table of available resources
+            - dict with mapping per partition {partition: {max_nprocs, max_ngpus, max_mem, ...}}
+        """
+        # Resources displayed in table of available resources (column labels in display order)
+        resources_display = ["Idle Cores", "Total Cores", "Total Nodes"]
```

Comment:
I'm wondering if it would be better to have this as a separate traitlet, that could be dict `{"displayed name": "resources_info key"}`.
This would separate the configuration of parsing from the one for displaying the output.
It would also allow to return a basic type (`dict`) here.

Reply:
I also wondered how to avoid having a separate list for the labels of available resources. The problem with dicts is keeping the order between the dict holding the labels and the dict of partitions with the counters.

Having a separated traitlet defining the displayed labels as you suggest seems good. Maybe we can do that, then remove `available_counts` from from `resources_info` and construct a separate object with just the table of available resources as a simple bi-dimensional list easy to template in jinja:
```
available_resources = [
    ["Partition", "Label 1", "Label 2", ...],
    [partition_name1, counter1, counter2, ...],
    [partition_name2, counter1, counter2, ...],
]
```